### DISCUSSION TITLE
Simple Window Switcher option: Alt-Tab only switches apps

### DISCUSSION THREAD

**Discussion by alexmcco** [valinet/ExplorerPatcher#665] [Ideas]
I would like an option in the Simple Window Switcher to make Alt-Tab only list [perhaps the frontmost window of] apps running, instead of all apps and all their windows. 

This is because I believe there is a certain cognitive load advantage to having the two different hotkeys (Alt-Tab, and Alt-`) execute only their specific and distinct functions: switching apps, and switching windows within the current app. 

Thank you to all contributors. ExplorerPatcher is a must-have for me on Windows, coming from MacOS, and trying to salvage my computing productivity.

**Comment by valinet** [comment]
I have an initial implementation in the latest pre-release (22000.434.41.6).

Idk how well it makes sense, but this is all I could come up with for now. Other than the actual usefulness (in Windows, `Alt`+`Tab` traditionally switches among all windows, I don't like it when only apps are shown as it is the default in some GNU/Linux desktop environments and maybe on OS X), the window previews were my biggest concern: for now, as you suggested, the preview of the last window of that application is shown. Maybe the design and layout should be better rearchitected to emphasize the icons more when in this mode... anyway, obtaining proper icons is kind of hard, there is no real, universal way to do it in Windows and not all apps may support/offer high resolution ones. For now, the current implementation will do.

There is an option in Properties - Window switcher to enable this: "Display only the most recent window (per application switcher)".

The settings are stored, as usual, in `HKCU\Software\ExplorerPatcher\sws`:

* `SwitcherIsPerApplication` (enables this mode)
* `AlwaysUseWindowTitleAndIcon` (in this mode, the displayed previews show the icon and the title of the executable file; if you set this to 1, it will instead show the icon and window title of the window whose preview is shown, as is the case for the regular, per-window `Alt`-`Tab`).

Beware, this mode might behave unexpectedly (at the moment, at least). I mean, idk what the expectations should be, really: what is the correct criteria for the things to display in this mode. Think about this, for example: on one hand, programs like web browsers open a new process for each new window. You'd want those grouped. So group by location of the executable that created the window. On the other, programs like `rundll32` execute arbitrary code from any location. Those can be totally different programs. You'd want those ungrouped, but the criteria above doesn't hold. Also, for a `rundll32`-like app, what should be displayed: each individual window, or windows created by an instance, or a single entry... it's actually more complicated to do this right than it seems on the surface. It kind of works for now, but idk, I don't think it is 100% correct. `explorer.exe` (for the taskbar) and Task Manager seem to do it properly. I personally don't know how to do it properly properly at the moment, neither have I looked how they decide how to group windows into apps, I am sure it is pretty complicated, needing to cover all the possibilities. I am too tired to disassemble and wonder around their implementation. Pity they never open up things like these, one always has to dig a lot. Doing it the same way as them would benefit both this mode and `Alt`-`; hopefully someone can contribute with a better implementation on this front, eventually a reimplementation of how Microsoft does it.

Thanks.

**Reply by alexmcco** [reply]
Thank you for your work and reply. I understand your concern with the complication of the implementation. I've been testing it today and I haven't run into any issues or unexpected behavior, from my perspective.

**Reply by valinet** [reply]
Thanks for the feedback. That’s good to hear. It turned rather neat, indeed, I have to admit.

**Reply by mukundshah** [reply]
> `AlwaysUseWindowTitleAndIcon` (in this mode, the displayed previews show the icon and the title of the executable file; if you set this to 1, it will instead show the icon and window title of the window whose preview is shown, as is the case for the regular, per-window `Alt`+`Tab`).

Can you please elaborate on this?
I tried this one, but didn't do anything. Was this supposed to make window switcher appear like this:
https://www.intego.com/mac-security-blog/wp-content/uploads/2017/06/master-the-app-switcher-1024x640.jpg
Only app icon and title?
I'm on version `22000.469.41.16`

**Reply by valinet** [reply]
No, no mode like that is implemented. When `SwitcherIsPerApplication` is on, the switcher displays previews of the last used window from each app. The title and icon for that preview are taken from the executable file of that respective application. If you set `AlwaysUseWindowTitleAndIcon`, rather than querying the executable directly, it firsts asks the window for the title and the icon, and then, if the window does not provide that window, it queries the executable. So yeah, it does something if you watch closely.

**Reply by mukundshah** [reply]
So, setting `AlwaysUseWindowTitleAndIcon` has any perfomance changes? Which is good, querying the executable directly or asking the window first?

**Reply by valinet** [reply]
With `SwitcherIsPerApplication` set to 1:

* `AlwaysUseWindowTitleAndIcon=0`:

![image](https://user-images.githubusercontent.com/6503598/156449696-c72363fd-b449-4796-af6f-2867d0836826.png)

* `AlwaysUseWindowTitleAndIcon=1`:

![image](https://user-images.githubusercontent.com/6503598/156449852-efb96aaa-91e2-48ad-ada2-025a3c1f6425.png)

As I said, `AlwaysUseWindowTitleAndIcon=1` displays the title of the latest window regardless whether multiple window of the program are open or not. Idk why I made that an option, but it's there, it's kind of confusing, but anyway, it's there for now.

When using sws in regular mode, with one preview per window, it queries each window first and then falls back to extracting the cion from the executable, so any mode you pick doesn't have any impact on performance compared to the stock use case, which performs very well, I may say. Gathering the window title is a non-blocking call basically, the system stores that information and can offer it instantaneously. For icons, one has to query the window, sending it a message - that can take time; because of that, the process is done asynchronously: the job is launched, and as the icons arrive, it notifies the main thread. When all icons are gathered, or there was a timeout obtaining them, a redraw operation is scheduled and the icons are drawn on the window, be it displayed already or not. So only 1 redraw is done in order to prioritize efficiency (I could schedule a redraw each time I obtained an icon, but in practice it's not worth it). This is done so that when it takes long to gather the icons, the switcher window can display without delay. I have spent quite some time optimizing this, as far as I know, there are no operations that can block the switcher for longer when displaying. Yeah, initially icons for windows might not display in some cases, but the switcher is there, can be used, it responds.

**Reply by JonasDoe** [reply]
Just stumbled upon the ExplorerPatcher, and really, this feature is a life-saver when you have, say, mutliple WebStorm, multiple GoLand and multiple Chrome Windows opened (amonst all the other clutter)! In Ubuntu `WIN`+\` allowed to toggle between instances, while `WIN`+`TAB` allowed to toggle between the (grouped) applications, and it's really great to find this feature in Windows as well. Thank you so much!
A minor question: Is it possible to change the keybinds so that `WIN`+`TAB` and `WIN`+\` are used instead of `ALT`+`WIN` and `ALT`+\`? This would render it a bit easier to switch between Windows and Ubuntu all the time.
Anyway, thank you again for this great tool! :)

**Reply by pyrates999** [reply]
> Just stumbled upon the ExplorerPatcher, and really, this feature is a life-saver when you have, say, mutliple WebStorm, multiple GoLand and multiple Chrome Windows opened (amonst all the other clutter)! In Ubuntu `WIN`+` allowed to toggle between instances, while `WIN`+`TAB` allowed to toggle between the (grouped) applications, and it's really great to find this feature in Windows as well. Thank you so much! A minor question: Is it possible to change the keybinds so that `WIN`+`TAB` and `WIN`+` are used instead of `ALT`+`WIN` and `ALT`+`? This would render it a bit easier to switch between Windows and Ubuntu all the time. Anyway, thank you again for this great tool! :)

Please submit this under a new discussion.

**Reply by JonasDoe** [reply]
> > Just stumbled upon the ExplorerPatcher, and really, this feature is a life-saver when you have, say, mutliple WebStorm, multiple GoLand and multiple Chrome Windows opened (amonst all the other clutter)! In Ubuntu `WIN`+`allowed to toggle between instances, while`WIN`+`TAB`allowed to toggle between the (grouped) applications, and it's really great to find this feature in Windows as well. Thank you so much! A minor question: Is it possible to change the keybinds so that`WIN`+`TAB`and`WIN`+` are used instead of `ALT`+`WIN` and `ALT`+`? This would render it a bit easier to switch between Windows and Ubuntu all the time. Anyway, thank you again for this great tool! :)
> 
> Please submit this under a new discussion.

As a question as "is it possible that ...", or in case it's not, as a suggestion?

**Reply by pyrates999** [reply]
all new features are to be posted under discussions.  And since your feature request is not what the original poster is asking for, it be easier if it was a new discussion.  You can link it to this one so that you can say they are related to each other.

**Reply by JonasDoe** [reply]
> all new features are to be posted under discussions. And since your feature request is not what the original poster is asking for, it be easier if it was a new discussion. You can link it to this one so that you can say they are related to each other.

Okay, just wasn't sure whether it acutally IS a new feature. Just created a new discussion for it.
